runtime/v2/shim_load.go

```diff
@@ -77,6 +77,8 @@ func (m *ShimManager) loadShims(ctx context.Context) error {
 		if len(id) > 0 && id[0] == '.' {
 			continue
 		}
+		logger := log.G(ctx).WithField("id", id)
```

Comment:
There's a "global" (for this function) ctx logger defined outside the loop, so I thought doing that would overwrite that one (in a loop), or perhaps I'm not thinking correctly 🥹

Reply:
True, `ctx := log.WithContext(ctx, log.G(ctx).WithField("id", id))` will work though

Reply:
Ah, hm.. right; so I now recall why I didn't want to use the context for that, becausee there's some _potential_ issues with that approach; when attaching the modified logger to the context; the extra field added will be inherited by _any_ code that gets passed the context and does logging. _Perhaps_ that's the intent, but it may not be at all times.

The second potential issue is additional overhead; the `WithField(s)` options are cumulative; adding additional field(s) means that a _new_ struct (copy) is created, which involves reflection, and merging the new fields; https://github.com/containerd/containerd/blob/a68efb1bad631a7a6c37953af791abf1d551790f/vendor/github.com/sirupsen/logrus/entry.go#L124-L151


There _may_ be some overhead in getting the logger from the context as well (instead of from a local var), but I have not measured that. And perhaps the overhead is not a concern (but I'm always a bit cautious especially with things running in a loop).


I just wrote a very basic benchmark to compare;

```
go test -v -bench .
goos: darwin
goarch: arm64
pkg: github.com/containerd/containerd/v2/foo
BenchmarkCtxLogger
BenchmarkCtxLogger-10       	      92	  11151721 ns/op	10755728 B/op	   55699 allocs/op
BenchmarkSharedLogger
BenchmarkSharedLogger-10    	     121	   9969444 ns/op	 9393327 B/op	   51689 allocs/op
PASS
ok  	github.com/containerd/containerd/v2/foo	3.316s
```

Code below:

<details>

```go
package foo

import (
	"context"
	"fmt"
	"io"
	"testing"

	"github.com/containerd/log"
)

func BenchmarkCtxLogger(b *testing.B) {
	b.ReportAllocs()
	ctx := context.Background()
	for i := 0; i < b.N; i++ {
		shimLoad(ctx)
	}
}

func BenchmarkSharedLogger(b *testing.B) {
	b.ReportAllocs()
	ctx := context.Background()
	for i := 0; i < b.N; i++ {
		shimLoadShared(ctx)
	}
}

func shimLoad(ctx context.Context) {
	log.G(ctx).Logger.Out = io.Discard

	ctx = log.WithLogger(ctx, log.G(ctx).WithField("namespace", "hello-world"))

	for i := 0; i < 1000; i++ {
		id := i
		ctx = log.WithLogger(ctx, log.G(ctx).WithFields(log.Fields{
			"id":      id,
			"field-a": "orig",
			"field-b": "orig",
			"field-c": "orig",
			"field-d": "orig",
			"field-e": "orig",
			"field-f": "orig",
		}))

		err := fmt.Errorf("some-error")
		log.G(ctx).WithError(err).Info("loop-da-loop")

		someFunc(ctx, id)
	}
}

func shimLoadShared(ctx context.Context) {
	log.G(ctx).Logger.Out = io.Discard

	ctx = log.WithLogger(ctx, log.G(ctx).WithField("namespace", "hello-world"))

	for i := 0; i < 1000; i++ {
		id := i
		logger := log.G(ctx).WithFields(log.Fields{
			"id":      id,
			"field-a": "orig",
			"field-b": "orig",
			"field-c": "orig",
			"field-d": "orig",
			"field-e": "orig",
			"field-f": "orig",
		})

		err := fmt.Errorf("some-error")
		logger.WithError(err).Info("loop-da-loop")

		someFunc(ctx, id)
	}
}

func someFunc(ctx context.Context, id int) {
	log.G(ctx).WithFields(log.Fields{
		"id":            id,
		"field-a":       "someFunc",
		"field-b":       "someFunc",
		"field-c":       "someFunc",
		"field-d":       "someFunc",
		"field-e":       "someFunc",
		"field-f":       "someFunc",
		"extra-field-a": "someFunc",
		"extra-field-b": "someFunc",
		"extra-field-c": "someFunc",
		"extra-field-d": "someFunc",
		"extra-field-e": "someFunc",
		"extra-field-f": "someFunc",
	}).Info("logging from someFunc")
}
```

</details>